File: backend/src/main/java/com/ufc/backend/backend/model/Classes.java
package com.ufc.backend.backend.model;

import com.ufc.backend.backend.model.based.FeedbackBased;
import lombok.*;
import org.springframework.data.annotation.Id;
import org.springframework.data.mongodb.core.mapping.DBRef;
import org.springframework.data.mongodb.core.mapping.Document;

import java.util.HashSet;
import java.util.List;
import java.util.Set;


@AllArgsConstructor
@NoArgsConstructor
@Getter
@Setter
@ToString
@Document
public class Classes extends FeedbackBased {
    @Id
    private String id;
    private String name;
    @DBRef
    private Classes preRequisite;
    private Integer hours;
    private Integer semester;
    @DBRef
    private Set<Teacher> teachers = new HashSet<>();

    @Override
    public int hashCode() {
        return super.hashCode();
    }

    @Override
    public boolean equals(Object obj) {
        if (getClass() != obj.getClass()) {
            return false;
        }
        return this.getId().equals(((Classes) obj).getId());
    }
}

File: backend/src/main/java/com/ufc/backend/backend/model/Classes.kt
package com.ufc.backend.backend.model

import com.fasterxml.jackson.annotation.JsonIgnore
import com.ufc.backend.backend.commons.model.Identifiable
import com.ufc.backend.backend.commons.model.FeedbackBased
import org.springframework.data.annotation.Id
import org.springframework.data.mongodb.core.mapping.DBRef
import org.springframework.data.mongodb.core.mapping.Document
import java.beans.ConstructorProperties

@Document
class Classes
@ConstructorProperties
    (
    "id",
    "name",
    "preRequisite",
    "hours",
    "semester",
    "teachers",
    "preRequisiteId",
)
constructor(
    @Id
    override var id: String,
    var name: String,
    @DBRef
    var preRequisite: Classes? = null,
    var hours: Int,
    var semester: Int,
    @DBRef
    var teachers: Set<Teacher>? = null,
) : Identifiable<String>, FeedbackBased() {

    override fun equals(other: Any?): Boolean {
        return if (javaClass != other!!.javaClass) {
            false
        } else this.id == (other as Classes).id
    }

    override fun toString(): String {
        return "Classes[id=$id, name=$name]"
    }

    override fun hashCode(): Int {
        return id.hashCode()
    }
}